File: rabbit-native-jvmti/src/main/java/com/susion/rabbit/jvmti/RabbitJvmTi.java
package com.susion.rabbit.jvmti;

import android.content.Context;
import android.os.Build;
import android.os.Debug;
import android.util.Log;

import java.io.File;
import java.lang.reflect.Method;
import java.nio.file.Files;
import java.nio.file.Paths;

/**
 * susionwang at 2020-04-24
 */
public class RabbitJvmTi {

    private static final String TAG = "RabbitJvmTi";
    private static final String RABBIT_JVMTI_AGENT = "rabbit-jvmti";


    public static void init(Context context, NativeCommunityHandler communityHandler) {
        attachJvmTiAgent(context);
        registerCommunityHandler(communityHandler);
    }

    private static void attachJvmTiAgent(Context context) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                ClassLoader classLoader = context.getClassLoader();
                Method findLibrary = ClassLoader.class.getDeclaredMethod("findLibrary", String.class);
                String jvmtiAgentLibPath = (String) findLibrary.invoke(classLoader, RABBIT_JVMTI_AGENT);
                Log.d(TAG, "jvmtiagentlibpath " + jvmtiAgentLibPath);

                File filesDir = context.getFilesDir();
                File jvmtiLibDir = new File(filesDir, "jvmti");
                if (!jvmtiLibDir.exists()) {
                    jvmtiLibDir.mkdirs();

                }
                File agentLibSo = new File(jvmtiLibDir, "agent.so");
                if (agentLibSo.exists()) {
                    agentLibSo.delete();
                }
                Files.copy(Paths.get(new File(jvmtiAgentLibPath).getAbsolutePath()), Paths.get((agentLibSo).getAbsolutePath()));

                Log.d(TAG, agentLibSo.getAbsolutePath() + "," + context.getPackageCodePath());

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    Debug.attachJvmtiAgent(agentLibSo.getAbsolutePath(), null, classLoader);
                    Log.d(TAG, "Debug.attachJvmTiAgent ----> ");
                } else {
                    try {
                        Class vmDebugClazz = Class.forName("dalvik.system.VMDebug");
                        Method attachAgentMethod = vmDebugClazz.getMethod("attachAgent", String.class);
                        attachAgentMethod.setAccessible(true);
                        attachAgentMethod.invoke(null, agentLibSo.getAbsolutePath());
                    } catch (Exception ex) {
                        ex.printStackTrace();
                    }
                }
                System.loadLibrary(RABBIT_JVMTI_AGENT);
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public native static void registerCommunityHandler(NativeCommunityHandler handler);

}

File: rabbit-native-jvmti/src/main/java/com/susion/jvmti/RabbitJvmTi.kt
package com.susion.jvmti

import android.content.Context
import android.os.Build
import android.os.Debug
import com.susion.rabbit.base.RabbitLog
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths

/**
 * susionwang at 2020-04-23
 */

object RabbitJvmTi {

    private val TAG = "RabbitJvmTi"
    private val RABBIT_JVMTI_AGENT = "rabbit-jvmti"

    fun init(context: Context) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val classLoader = context.classLoader
                val findLibrary =
                    ClassLoader::class.java.getDeclaredMethod("findLibrary", String::class.java)
                val jvmtiAgentLibPath =
                    findLibrary.invoke(classLoader, RABBIT_JVMTI_AGENT) as String
                RabbitLog.d(TAG, "jvmtiagentlibpath: $jvmtiAgentLibPath")
                val filesDir = context.filesDir
                val jvmtiLibDir = File(filesDir, "jvmti")
                if (!jvmtiLibDir.exists()) {
                    jvmtiLibDir.mkdirs()
                }

                val agentLibSo = File(jvmtiLibDir, "agent.so")
                if (agentLibSo.exists()) {
                    agentLibSo.delete()
                }
                Files.copy(
                    Paths.get(File(jvmtiAgentLibPath).absolutePath),
                    Paths.get(agentLibSo.absolutePath)
                )

                RabbitLog.d(TAG, agentLibSo.absolutePath + "," + context.packageCodePath)

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    Debug.attachJvmtiAgent(agentLibSo.absolutePath, null, classLoader)
                } else {
                    try {
                        val vmDebugClazz = Class.forName("dalvik.system.VMDebug")
                        val attachAgentMethod =
                            vmDebugClazz.getMethod("attachAgent", String::class.java)
                        attachAgentMethod.isAccessible = true
                        attachAgentMethod.invoke(null, agentLibSo.absolutePath)
                    } catch (ex: Exception) {
                        ex.printStackTrace()
                    }

                }
                System.loadLibrary(RABBIT_JVMTI_AGENT)
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

}